### PR TITLE
Feature/upstream by state

### PR DESCRIPTION
**Pull request by romsharon98**
related: https://github.com/apache/airflow/pull/41506
change the target branch from airflow3 to 2.10 according to [this comment ](https://github.com/apache/airflow/pull/41506#issuecomment-2295402432) on the related PR.

this PR add an ability to get all upstream tasks by state.
the current state is that user can get only upstream tasks without any state.

this can solve issues like this https://stackoverflow.com/questions/73740427/airflow-how-to-get-list-of-upstream-failed-tasks

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/models/taskinstance.py
Comment (uranusjr) [inline airflow/models/taskinstance.py:4043]: It’s probbly better to do this the other way around? i.e. get the upstream task IDs first, and then filter them by state. The performance difference is likely significant if the DAG is pretty large (and `get_task_instances` may need to fetch a lot more rows than needed).

**Comment by jedcunningham** [comment]
I'm not sure we should be adding this to 2.x at this point - this is not a bugfix or something appropriate for a bridge release imo. Even if we did, this should go to `v2-10-test`, not stable.

**Review by ephraimbuddy — CHANGES_REQUESTED** [review]
Blocking this PR since we have a resolution of no new features in Airflow 2. I am happy to unblock it if we decide to have it in

**Pull request closed without merging**
